Add a guide for external network use case

mkimuram opened the pull request:
fixes: #572

mkimuram committed:
Add a guide for external network use case

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram commented on src/content/getting-started/quickstart/external/_index.md:
````diff
@@ -0,0 +1,429 @@
+```bash
+export KUBECONFIG=kubeconfig.cluster-a
+kubectl -n default run tmp-shell --rm -i --tty --image quay.io/submariner/nettest -- bash
+curl 242.0.255.253
````
Due to this [issue](https://github.com/submariner-io/lighthouse/issues/603), access from cluster to non-cluster works only via global ingress IP. It should be able to be accessed via DNS name.

mkimuram commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,429 @@
+metadata:
+  name: test-vm
+spec:
+  ports:
```
Does it make sense to make it headless service, by adding `clusterIP: None`?
Then, it will be headless service without selector.

By doing so, I expect that it can provide a persistent source IP from the endpoint that is pointing to an external host.
Currently, it doesn't seem to work maybe the similar reason to this [issue](https://github.com/submariner-io/lighthouse/issues/603).

mkimuram replied:
Opened a separate issue https://github.com/submariner-io/submariner/issues/1537 .

mkimuram commented on src/content/getting-started/quickstart/external/_index.md:
````diff
@@ -0,0 +1,429 @@
+subctl join --kubeconfig kubeconfig.cluster-b broker-info.subm --clusterid cluster-b --natt=false
+```
+
+#### Deploy dns server on cluster-a for non-cluster hosts
````
Created a separate issue to discuss a better way for this, [here](https://github.com/submariner-io/lighthouse/issues/600).

mkimuram commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,429 @@
+    On the other hand, subnets of non-cluster hosts should be distinguished from those of all the clusters to easily specify the external network CIDR.
+    In above case, cluster-a and cluster-b belongs to 192.168.122.0/25 network and test-vm belongs to 192.168.122.128/25 network.
+    Therefore, the external network CIDR for this configuration is 192.168.122.128/25.
+    In test environment for just one host, we will be able to specify external network CIDR, like 192.168.122.142/32, however design of the subnet needs to be considered when more hosts are added.
```
For this configuration to be more flexible, it would be better to provide a separate interface for external CIDR, in the future, as @mangelajo has already pointed out [here](https://docs.google.com/document/d/1knqUrD5kKI0u8X7dRUf5IseYa-13BKSRtry9ulD-DDQ/edit?pli=1#bookmark=id.gr1uxfycb496).

mkimuram reviewed (commented):
Commenting on what seems to be missing. Beside them, it seems working well.

mkimuram commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+#### Install `subctl` on node of cluster-a
+
+{{% subctl-install %}}
+
```
Just to be sure, currently tested with `curl -Ls https://get.submariner.io | VERSION="devel" bash -s -`, until globalnet v2 is enabled on released version.

sridhargaddam replied:
0.10 is now released :)

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+---
+
+This guide covers the minimum steps to try external network use case.
+In this use case, users can access between pods in clusters and applications outside clusters
+by using DNS names that lighthouse provides or globalnet ingress IPs.
+In addition, source IPs of these communications can be preserved.
```
```suggestion
In this use case, pods running in the K8s cluster can access external applications outside of the cluster and vice-versa by using DNS resolution supported by lighthouse or manually using the globalnet ingress IPs.
Along with connectivity, the source IP of the traffic will also be preserved.
```

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+by using DNS names that lighthouse provides or globalnet ingress IPs.
+In addition, source IPs of these communications can be preserved.
+
+This feature is still experimental, so the behaviors and configurations may be changed in the future.
```
```suggestion
This feature is still experimental, so the configuration mechanism and observed behavior may be changed in the future.
```

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+
+1. Prepare below:
+    - Two or more k8s clusters
+    - One or more non-cluster hosts, that exist in the same network segment to the one of the k8s clusters
```
```suggestion
    - One or more non-cluster hosts, that exist in the same network segment to one of the k8s clusters
```

sridhargaddam reviewed (changes requested):
Thanks @mkimuram for the documentation.

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+    In above case, all of them are deployed inside 192.168.122.0/24 segment.
+    However, it is only required that cluster-a and test-vm are in the same segment.
+    Other clusters, cluster-b and any additional clusters, can be deployed in different segments or even in any other networks in the internet.
+    Also, these clusters can be multi-node cluster.
```
```suggestion
    Also, these clusters can be multi-node clusters.
```

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+    | cluster-a |10.42.0.0/24  |10.43.0.0/16  |
+    | cluster-b |10.42.0.0/24  |10.43.0.0/16  |
+
+    Note that we will use globalnet in this guide, therefore overlapping CIDR is allowed.
```
```suggestion
    Note that we will use globalnet in this guide, therefore overlapping CIDRs are supported.
```

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+    [here](https://submariner.io/getting-started/quickstart/k3s/) until "Deploy cluster-b on node-b",
+    with modifying deploy commands to just `curl -sfL https://get.k3s.io | sh -` to use default CIDR.
+
+### Set up submariner
```
```suggestion
### Setup Submariner
```

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
````diff
@@ -0,0 +1,453 @@
+
+```bash
+subm_gw_ip=192.168.122.26
+ip r add 242.0.0.0/8 via ${subm_gw_ip}
````
Do you want to add a small note saying 242.0.0.0/8 is the default globalCIDR?

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
````diff
@@ -0,0 +1,453 @@
+python -m http.server 80
+```
+
+##### Verify access to hosts from clusters
````
```suggestion
##### Verify access to External hosts from clusters
```

sridhargaddam commented on src/content/getting-started/quickstart/external/_index.md:
```diff
@@ -0,0 +1,453 @@
+
+##### Verify access to hosts from clusters
+
+Create Service, Endpoints, ServiceExport to access to test-vm, on cluster-a:
```
```suggestion
Create Service, Endpoints, ServiceExport to access the test-vm from cluster-a:
```

mkimuram commented:
@sridhargaddam

Thank you for your review and suggestions! Fixed as suggested.

mangelajo commented on src/content/getting-started/quickstart/external/_index.md:
````diff
@@ -0,0 +1,454 @@
+```bash
+CLUSTER_CIDR=10.42.0.0/24
+EXTERNAL_CIDR=192.168.122.128/25
+subctl join --kubeconfig kubeconfig.cluster-a broker-info.subm --clusterid cluster-a --natt=false --clustercidr=${CLUSTER_CIDR},${EXTERNAL_CIDR}
````
```suggestion
subctl join --kubeconfig kubeconfig.cluster-a broker-info.subm --clusterid cluster-a --natt=false --external-network-cidr ${EXTERNAL_CIDR}
```

mangelajo replied:
In the future we could have something like this, but still not implemented

mkimuram commented:
Still working.

mkimuram commented:
Updated to add notes for missing features and configuration consideration.

nerdalert commented:
Hi @mkimuram, I was wondering if you would be up for chatting about this proposal as there is some interest in a project I am working on to find a bare metal and/or non-kube solution and this PR was referenced. My email is brent.salisbury@redhat.com if you were up to chatting on slack/hangout or whatever would be convenient for you. Thanks!

nyechiel commented:
bump

nyechiel commented:
@dfarrell07 can you please review this one for style/grammar? I don't see a reason not to merge it, as the content is accurate AFAIK and we keep point users to this PR instead of a proper doc. Thanks!

@sridhargaddam FYI

dfarrell07 commented:
Is this ready to review/merge, the relevant implementation is done?

dfarrell07 commented:
> @dfarrell07 can you please review this one for style/grammar? I don't see a reason not to merge it, as the content is accurate AFAIK and we keep point users to this PR instead of a proper doc. Thanks!

ACK, sounds good @nyechiel will do. I thought this was waiting on some still-pending implementation PRs.

dfarrell07 commented:
I'll send a commit with a bunch of little tweaks if that's okay.

dfarrell07 commented:
> I'll send a commit with a bunch of little tweaks if that's okay.

Or maybe it would be easier to merge this as-is and I can send a follow-up PR (https://github.com/submariner-io/submariner-website/pull/633/commits/3aabdcd4268a266394327dbee3105326472b10f6) with some refactoring?

We should be sure to squash+merge, not rebase+merge.

Thanks @mkimuram!

mkimuram committed:
Edit external network quickstart guide

Various edits to the new quickstart guide for connecting K8s clusters
with non-clusters using Submariner.

Signed-off-by: Daniel Farrell <dfarrell@redhat.com>
Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram commented:
@dfarrell07   

Thank you for your review and fixes.
I've rebased and squashed my commits and added your commit. PTAL

sridhargaddam reviewed (approved):
Thank you for your patience on the PR @mkimuram. As discussed on the thread, we can get this in and iterate using further PRs.
